### DISCUSSION TITLE
Freezable/Immutable  property

### DISCUSSION THREAD

**Discussion by CodingOctocat** [dotnet/csharplang#7669] [Language Ideas]
I don't know how implement the "Freezable/Immutable" property, but it would be simple to add a language feature that would accomplish this:

```csharp
public class MyClass
{
    public Foo MyFoo { get; freez; }

    public void Do()
    {
        MyFoo.Property1 = xxx; // Ok
    }
}

var cls = new MyClass();
cls.MyFoo = new MyFoo(); // Error
cls.MyFoo.Property1 = xxx; // Error
cls.MyFoo.Property1.SubProperty1 = xxx; // Error
cls.MyFoo.ListProperty[0] = xxx; // Error
```

Possible syntax:

```csharp
public Foo MyFoo { get; freez; }
public freez Foo MyFoo { get; }

public Foo MyFoo { get; private freez;} // Debatable
public freez Foo MyFoo { get; private set; }

// Filed support
public freez Foo MyFoo;
```

**Comment by HaloFour** [comment]
Don't add a `set` accessor, or make it an `init` accessor if you want it mutable during initialization.

**Reply by huoyaoyuan** [reply]
This does not provide deep immutability though.

**Reply by HaloFour** [reply]
No, but I don't see how this proposal would either.  A keyword on a property wouldn't prevent you from passing it's value by reference to some arbitrary method that would be totally unaware of that restriction.

**Reply by CodingOctocat** [reply]
Perhaps only deep clones can be used at this stage? Is there any other more elegant way, such as a proxy like `Freezable<Foo>/Immutable<Foo>`.

```csharp
private Foo _myFoo;
public Foo MyFoo => _myFoo.DeepClone()
```
